pkg/services/apiserver/auth/authorizer/org_id.go

```diff
@@ -1,87 +0,0 @@
-package authorizer
-
-import (
-	"context"
-	"fmt"
-
-	"k8s.io/apiserver/pkg/authorization/authorizer"
-
-	claims "github.com/grafana/authlib/types"
-	"github.com/grafana/grafana/pkg/apimachinery/identity"
-	"github.com/grafana/grafana/pkg/infra/log"
-	"github.com/grafana/grafana/pkg/services/org"
-)
-
-var _ authorizer.Authorizer = &orgIDAuthorizer{}
-
-type orgIDAuthorizer struct {
-	log log.Logger
-	org org.Service
-}
-
-func newOrgIDAuthorizer(orgService org.Service) *orgIDAuthorizer {
-	return &orgIDAuthorizer{
-		log: log.New("grafana-apiserver.authorizer.orgid"),
-		org: orgService,
-	}
-}
-
-func (auth orgIDAuthorizer) Authorize(ctx context.Context, a authorizer.Attributes) (authorized authorizer.Decision, reason string, err error) {
-	signedInUser, err := identity.GetRequester(ctx)
-	if err != nil {
-		return authorizer.DecisionDeny, fmt.Sprintf("error getting signed in user: %v", err), nil
-	}
-
-	info, err := claims.ParseNamespace(a.GetNamespace())
-	if err != nil {
-		return authorizer.DecisionDeny, fmt.Sprintf("error reading namespace: %v", err), nil
-	}
-
-	// No opinion when the namespace is empty
-	if info.Value == "" {
-		return authorizer.DecisionNoOpinion, "", nil
-	}
-
-	// Grafana super admins can see things in every org
-	if signedInUser.GetIsGrafanaAdmin() {
-		return authorizer.DecisionNoOpinion, "", nil
-	}
-
-	if info.OrgID == -1 {
-		return authorizer.DecisionDeny, "org id is required", nil
-	}
-
-	if info.StackID != 0 {
-		return authorizer.DecisionDeny, "using a stack namespace requires deployment with a fixed stack id", nil
-	}
-
-	// Quick check that the same org is used
-	if signedInUser.GetOrgID() == info.OrgID {
-		return authorizer.DecisionNoOpinion, "", nil
-	}
-
-	// If we have an anonymous user, let the next authorizers decide.
-	if signedInUser.GetIdentityType() == claims.TypeAnonymous {
-		return authorizer.DecisionNoOpinion, "", nil
-	}
-
-	// Check if the user has access to the specified org
-	// nolint:staticcheck
-	userId, err := signedInUser.GetInternalID()
-	if err != nil {
-		return authorizer.DecisionDeny, "unable to get userId", err
-	}
-	query := org.GetUserOrgListQuery{UserID: userId}
-	result, err := auth.org.GetUserOrgList(ctx, &query)
```

Comment:
how does the new authorizer know if the user is in another org?   When i last checked (with multi-org setup) changing from `default` (org-1) to `org-2` would have the same user identity with only a single org

Reply:
So this is verified by authentication before we get here. But I see the problem. For you to authorize correctly in another org it would not be enough right now to change the url. This authorizer would always deny you if you don't set [X-Grafana-Org-Id](https://github.com/grafana/grafana/blob/main/pkg/services/authn/authnimpl/service.go#L494-L506) to match the org with the namespace.

I will think about this a bit but I am pretty sure we can solve it during authentication too.

Reply:
Well I have a bunch of failing tests now so need to look into why